common/include/pcl/common/impl/centroid.hpp

```diff
@@ -659,6 +660,270 @@ computeMeanAndCovarianceMatrix (const pcl::PointCloud<PointT> &cloud,
 }
 
 
+
+
+
+template <typename PointT, typename Scalar> inline unsigned int
+computeCentroidAndOBB (const pcl::PointCloud<PointT> &cloud,
+                  Eigen::Matrix<Scalar, 3, 1> &centroid,
+                  Eigen::Matrix<Scalar, 3, 1> &obb_position,
+                  Eigen::Matrix<Scalar, 3, 1> &obb_dimensions,
+                  Eigen::Matrix<Scalar, 3, 3> &obb_rotational_matrix)
+{
+    Eigen::Matrix<Scalar, 3, 3> covariance_matrix;
+    Eigen::Matrix<Scalar, 4, 1> centroid4;
+    unsigned int point_count= computeMeanAndCovarianceMatrix(cloud, covariance_matrix, centroid4);
+    if (!point_count)
+      return (0);
+    centroid(0) = centroid4(0);
+    centroid(1) = centroid4(1);
+    centroid(2) = centroid4(2);
+
+    Eigen::SelfAdjointEigenSolver<Eigen::Matrix<Scalar, 3, 3>> evd(covariance_matrix);
+
+    //Eigen::Vector<Scalar, 3> eigenvalues_ = evd.eigenvalues();
+    //major_value =    eigenvalues_(2); //promem order
+    //middle_value =   eigenvalues_(1);
+    //minor_value =    eigenvalues_(0);
+
+    Eigen::Matrix<Scalar, 3, 3> eigenvectors_ = evd.eigenvectors();
+    Eigen::Matrix<Scalar, 3, 1> major_axis;
+    Eigen::Matrix<Scalar, 3, 1> middle_axis;
+    Eigen::Matrix<Scalar, 3, 1> minor_axis;
+
+    minor_axis = eigenvectors_.col(0);//the eigenvectors do not need to be normalized (they are already)
+    middle_axis = eigenvectors_.col(1);
+
+    // cross product of the other two: ux = uy X uz   major = middle X minor
+    major_axis(0) = middle_axis(1) * minor_axis(2) - middle_axis(2) * minor_axis(1);
+    major_axis(1) = middle_axis(2) * minor_axis(0) - middle_axis(0) * minor_axis(2);
+    major_axis(2) = middle_axis(0) * minor_axis(1) - middle_axis(1) * minor_axis(0);
```

Comment:
Is it really necessary to do this? Are the eigenvectors from `SelfAdjointEigenSolver` not guaranteed to be orthogonal?

Reply:
i am not sure if they might be not always right hand rule, to be checked